### PR TITLE
#160 let's get rid of List.add() - Part 1

### PR DESCRIPTION
**Pull request by super132**
For task #160, replacing List.add() usage with Concat. This PR is split from #183

### CODE REVIEW

**Comment by super132** [comment]
@yegor256 , this is the basic implementation of `Concat` to get rid of `List.add()`.

**Comment by super132** [comment]
@davvd , This PR is split from #183 , not sure if it needs to be reviewed by architect again. Please advise. Thanks.

**Comment by davvd** [comment]
@super132 Thanks for the pull request, let me find a reviewer..

**Comment by davvd** [comment]
@krzyk please review, thanks

File: src/main/java/org/takes/misc/Concat.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    /**
+     * The concat iterator to traverse the input iterables as if they are
+     * from one list.
+     *
+     * <p>This class is NOT thread-safe.
+     */
+    private static class ConcatIterator<E> implements Iterator<E> {
+
+        /**
+         * Internal reference for holding the first iterator form constructor.
```
Comment (krzyk) [inline src/main/java/org/takes/misc/Concat.java]: @super132 `form` -> `from`

File: src/main/java/org/takes/misc/Concat.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+         */
+        private final transient Iterator<E> left;
+
+        /**
+         * Internal reference for holding the second iterator form constructor.
```
Comment (krzyk) [inline src/main/java/org/takes/misc/Concat.java]: @super132 `form` -> `from`

File: src/main/java/org/takes/misc/Concat.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+         */
+        private final transient Iterator<E> right;
+
+        /**
+         * Ctor. ConcatIterator traverses the element.
```
Comment (krzyk) [inline src/main/java/org/takes/misc/Concat.java]: @super132 no need to repeat the description of the class here, so just `Constructor.` is enough

File: src/main/java/org/takes/misc/Concat.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+package org.takes.misc;
+
+import java.util.Iterator;
+
+/**
+ * Concat iterable.
+ *
+ * @author Jason Wong (super132j@yahoo.com)
+ * @version $Id$
+ * @since 0.13.8
```
Comment (krzyk) [inline src/main/java/org/takes/misc/Concat.java]: @super132 should be `0.15.2`

File: src/test/java/org/takes/misc/ConcatTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+/**
+ * Test case for {@link Concat}.
+ *
+ * @author Jason Wong (super132j@yahoo.com)
+ * @version $Id$
+ * @since 0.13.8
```
Comment (krzyk) [inline src/test/java/org/takes/misc/ConcatTest.java]: @super132 should be `0.15.2`

File: src/test/java/org/takes/misc/ConcatTest.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+ */
+public final class ConcatTest {
+
+    /**
+     * Concat can concat.
```
Comment (krzyk) [inline src/test/java/org/takes/misc/ConcatTest.java]: @super132 *Concat can concatenate*

File: src/test/java/org/takes/misc/ConcatTest.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    @Test
+    public void concat() {
```
Comment (krzyk) [inline src/test/java/org/takes/misc/ConcatTest.java]: @super132 `concatenates()`

File: src/test/java/org/takes/misc/ConcatTest.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        blist.add(bone);
+        blist.add(btwo);
+        MatcherAssert.assertThat(
+                (Iterable<String>) new Concat<String>(alist, blist),
+                Matchers.hasItems(aone, atwo, bone, btwo)
+        );
+    }
+
+    /**
+     * Concat can concat with empty list.
```
Comment (krzyk) [inline src/test/java/org/takes/misc/ConcatTest.java]: @super132 *Concat can concatenate with empty list.*

File: src/test/java/org/takes/misc/ConcatTest.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    @Test
+    public void concatWithEmptyList() {
```
Comment (krzyk) [inline src/test/java/org/takes/misc/ConcatTest.java]: @super132 `concatenatesWithEmptyList()`

**Comment by krzyk** [comment]
@super132 please see my comments

**Comment by super132** [comment]
@krzyk, all comments are addressed. Please have a check. Thanks.

File: src/main/java/org/takes/misc/Concat.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+     */
+    private final transient Iterable<T> left;
+
+    /**
+     * Internal reference to hold the second elements from constructor.
+     */
+    private final transient Iterable<T> right;
+
+    /**
+     * To produce an iterable collection combining a and b, with order of the
```
Comment (krzyk) [inline src/main/java/org/takes/misc/Concat.java]: @super132 this description should be in the class description, here just `Constructor.`

**Comment by krzyk** [comment]
@super132 one more comment, and it looks like the build is failing

**Comment by krzyk** [comment]
@super132 failing build is not your fault, we will have to wait for #237

**Comment by super132** [comment]
@krzyk, comment is addressed and the build is fixed. Please check. Thanks.

**Comment by super132** [comment]
@krzyk , one of the travis CI build failing due to unable to lookup a maven dependency for qulice.

**Comment by krzyk** [comment]
@rultor merge pls

File: src/test/java/org/takes/misc/ConcatTest.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        final String atwo = "a2";
+        alist.add(aone);
+        alist.add(atwo);
+        final List<String> blist = new ArrayList<String>(2);
+        final String bone = "b1";
+        final String btwo = "b2";
+        blist.add(bone);
+        blist.add(btwo);
+        MatcherAssert.assertThat(
+            new Concat<String>(alist, blist),
```
Comment (yegor256) [inline src/test/java/org/takes/misc/ConcatTest.java]: I would rather do this:

```
assertThat(
  Joiner.on(" ").join(
    new Concat<String>(
      Arrays.asList("one", "two"),
      Arrays.asList("three", "four")
    )
  )
  Matchers.equalTo("one two three four")
);
```

looks more readable, isn't it?

**Comment by yegor256** [comment]
@super132 looks good, one comment from me above. tests don't look readable now, they are too long, see my suggestion

File: src/test/java/org/takes/misc/ConcatTest.java (hunk 11 above)
Comment (super132) [inline src/test/java/org/takes/misc/ConcatTest.java]: @yegor256, yes. It looks much simpler. I will modify accordingly.

**Comment by super132** [comment]
@yegor256, I have updated the codes per your suggestions. Please have a look. Thanks.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by super132** [comment]
@yegor256 , I have merged the latest source from master and hope it solves the merge conflict. Thanks.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by yegor256** [comment]
@rultor try to merge again

**Comment by super132** [comment]
@krzyk I saw pull requests can be merged now. Shall we try merging again? Thanks.

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by yegor256** [comment]
@rultor merge

**Comment by davvd** [comment]
@elenavolokhova review this ticket please, for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by elenavolokhova** [comment]
@krzyk 
>yegor256 added a note 7 days ago
I would rather do this: ...
looks more readable, isn't it?

This comment wasn't addressed. Please pay attention to all issues in the ticket.

**Comment by krzyk** [comment]
@yegor256 please add `@super132` in comment https://github.com/yegor256/takes/pull/232#discussion_r29545495

**Comment by davvd** [comment]
@rultor deploy pls

**Comment by yegor256** [comment]
@krzyk I can't find my comment... @elenavolokhova but I confirm that it was done wrong, I'll try to be more attentive in the future

**Comment by elenavolokhova** [comment]
@yegor256 I mean inline comments in files changed by current PR. Thanks :)

**Comment by elenavolokhova** [comment]
@davvd Quality is acceptable.

**Comment by davvd** [comment]
> @davvd Quality is acceptable.

@elenavolokhova thanks a lot, next time everybody should try to make it better

**Comment by davvd** [comment]
@krzyk I added 10 mins to @elenavolokhova (for QA review) in transaction 57166968

I just added **21 mins** to your account, many thanks for your contribution (57166973).. 153 hours and 25 mins spent here

review comments (c=6) added as a bonus

+21 added to your rating, at the moment it is: [+4833](http://www.netbout.com/b/35035?open=rating)

**Pull request merged**